runtime/decimal.c

```diff
@@ -3,49 +3,59 @@
 
 typedef GC decQuad *DecimalPtr;
 
-typedef struct {
-    TaggedPtr ptr;
-    PanicCode panicCode;
-} TaggedPtrPanicCode;
-
 // Status flags that we should treat as failure
 // XXX need to investigate DEC_Underflow
 #define DECIMAL_STATUS_FAIL (DEC_Errors|DEC_Subnormal)
 
-static inline PanicCode statusPanicCode(uint32_t status) {
-    if (status & DECIMAL_STATUS_FAIL) {
-        // XXX decide what panic codes to use 
-        return (status & DEC_Division_by_zero) ? PANIC_DIVIDE_BY_ZERO : PANIC_ARITHMETIC_OVERFLOW;
-    }
-    return 0;
-}
-
 static inline TaggedPtr createDecimal(decQuad *dq) {
     DecimalPtr dp = _bal_alloc(sizeof(decQuad));
     *dp = *dq;
-    return ptrAddFlags(dp, (uint64_t)TAG_DECIMAL);
+    return ptrAddFlags(dp, (uint64_t)TAG_DECIMAL << TAG_SHIFT);
+}
+
+static inline TaggedPtrPanicCode finish(decQuad *dq, decContext *cx) {
+    TaggedPtrPanicCode result;
+    uint32_t status = cx->status;
+    if (status & DECIMAL_STATUS_FAIL) {
+        // XXX decide what panic codes to use 
+        result.panicCode = (status & DEC_Division_by_zero) ? PANIC_DIVIDE_BY_ZERO : PANIC_ARITHMETIC_OVERFLOW;
+    }
+    else {
+        result.panicCode = 0;
+    }
+    if (result.panicCode == 0) {
```

Comment:
this if needs to merge with above if-else

Reply:
Fixed